### DISCUSSION TITLE
Difficulty in converting 32bit raw

### DISCUSSION THREAD

**Discussion by Diotsp** [ImageMagick/ImageMagick#4866] [Help] [answered]
Version: `7.1.0-24 Q16-HDRI x64`

I have a set of PNGs that I use Photoshop to change to 32bit and save as headerless RAWs.
![W](https://user-images.githubusercontent.com/100128498/154938743-c1e78f2c-1641-44ef-b5cc-eaf6a07133e0.png)
When I attempt to convert these back into a standard format using
 `convert -size 1100x823 -depth 32 -endian msb RGB:W32I.raw Wout.png`
 the results are highly distorted.
![Wout](https://user-images.githubusercontent.com/100128498/154938841-d9639519-175e-482f-8cae-91a3be591695.png)
I have tried several parameters but none seem to rectify the issue. However I have noticed that the divergence is much lesser at 16bit, and absent at 8.

**Comment by urban-warrior** [comment]
Start with a sanity check:

```
$ magick tree.png -depth 32 -endian msb RGB:W32I.raw
$ magick -size 1100x823 -depth 32 -endian msb RGB:W32I.raw Wout.png
$ magick compare -metric rmse Wout.png tree.png diff.png
0 (0)
```

A distortion of 0 means the image survived the round trip from PNG to 32-bit raw back to PNG.

**Reply by Diotsp** [reply]
`0 (0)`, with this for `diff.png`:
![diff](https://user-images.githubusercontent.com/100128498/155038828-d93e4278-cc70-452d-bdfb-03bfe978f061.png)

**Reply by fmw42** [reply]
What does this mean?  are you satisfied with the result or not?

**Comment by snibgo** [comment]
> I use Photoshop to change to 32bit

32-bits per channel per pixel, I assume. Unsigned integers, or signed integers, or IEEE floating-point, or what?

**Reply by Diotsp** [reply]
'32-bits/Channel' is how the mode is labelled, and I infer per-pixel. I cannot find any information on unsigned/signed/etc on the official help page:
[https://helpx.adobe.com/photoshop/using/file-formats.html](url)
Nor anywhere else I look.

**Reply by fmw42** [reply]
As I wrote in my example, it appears to be floating-point, not signed or unsigned!

**Comment by fmw42** [comment]
I get this after creating x.raw as 32-bit raw in PS:

`magick -size 1100x823 -depth 32 -interlace plane -endian msb -define quantum:format=floating-point RGB:x.raw -alpha off y.png`

![y](https://user-images.githubusercontent.com/13458218/155015119-05cd536d-c558-469c-ae23-9ab9034b8ed3.png)

I get the same looking result if I save it as 32-bit TIFF in PS and then convert to PNG in Imagemagick.  looks like a profile  difference as the profile is likely removed for raw data.

**Reply by Diotsp** [reply]
The '-define quantum:format=floating-point' also produces that result for me, which is much closer although notably darker to the original. I checked this is true regardless of interlacing.
When I use a 32-bit TIFF as you suggested, and `convert z.tif z.png`, I get the following error:
`convert: profile 'icc': 0h: PCS illuminant is not D50 'z.png' @ warning/png.c\MagickPNGWarningHandler\1748`
And the result is the same darker form. Using `compare -metric rmse` gives `0`, suggesting they are the same, but produces the same warning.

If this is the profile difference, is there any way to compensate this in IM?

**Reply by fmw42** [reply]
Your raw file does not have the sRGB profile that you have automatically with PNG. So the looks may differ. Check the actual pixel difference with "compare"

**Reply by Diotsp** [reply]
This is the output of `compare y.png z.png`
![dif](https://user-images.githubusercontent.com/100128498/155042580-502ae060-ccf6-4170-8a9e-0a46b5bfd070.png)
Which seems the same as that produced by the earlier 'sanity check'.

**Reply by fmw42** [reply]
Visual compare is not as good as the numerical value from say -metric rmse.

I do not know what PS does in the way of processing when converting to 32-bits/channel.

**Reply by Diotsp** [reply]
Comparing the two pngs, one from the .raw and one from the .tif, using `-metric rmse`, gives the numerical output `0 (0)`, followed by the aforementioned warning.

**Reply by fmw42** [reply]
That means the data is identical. So the difference in viewing has to do with profiles or other things. Look at the two identify -verbose listings and see what is different?

**Reply by Diotsp** [reply]
The `z.png` from the .tif has a `Profiles` section:
```
Profiles-8bim: 10524 bytes
Profile-icc: 516 bytes
Profile-iptc: 15 bytes
   City[1,90]: 0x00000000: 254700 -%G
   unknown[2,0]:
Profile-xmp: 16196 bytes
```
and under `Properties`, it has these additional lines:
```
dc:format: image/tiff
exif:ColorSpace: 65535
exif:PixelXDimension: 1100
exif:PixelYDimension: 823
icc:copyright: Copyright 2022 Adobe Systems Incorporated
icc:description: Adobe RGB (1998) (Linear RGB Profile)
photoshop:ColorMode: 3
photoshop:ICCProfile: Adobe RGB (1998) (Linear RGB Profile)
png:iCCP: chunk was found
png:pHYs: x_res=2834, y_res=2834, units=1
png:text-encoded profiles: 4 were found
quantum:format: floating-point
tiff:alpha:unspecified
tiff:endian: lsb
tiff: photometric: RGB
tiff:rows-per-strip: 79
tiff:software: Adbove Photoshop CC 2015 (Windows)
tiff:timestamp: 2022:02:21 16:56:18
```
It also has some xmp lines that are just timestamps or document IDs. I noticed two `Properties` lines were different:
`png:IHDR.interlace_method` is `1 (Adam7)` in y.png, and `0 (Not interlaced)` in z.png.
'png:text' is '2` in y.png, and `29` in z.png.
`Identify -verbose z.png` also produces the same ``profile 'icc': 0h: PCS illuminant is not D50 'z.png' @ warning/png.c\MagickPNGWarningHandler\1748`` warning.

**Reply by fmw42** [reply]
This has Adobe RGB (1998) (Linear RGB Profile) profile. Your input probably has no profile or sRGB profile. check that.

**Reply by fmw42** [reply]
A linear RGB profile is generally darker than a non-linear sRGB profile.

**Reply by Diotsp** [reply]
My original source png gives `Colorspace:sRGB` towards the top of the readout; the same as the z.png converted from z.tif, and the y.png converted from x.raw. However the y.png has no `icc:description: Adbove RGB` or `photoshop:ICCProfile:Adobe RGB` line like the other two.
Edit: to clarify, in the `Properties` section

**Comment by snibgo** [comment]
@Diotsp: I suggest you make a "raw" image in PS and paste a link here. Tell us how many channels it has, and the dimensions in pixels.

**Reply by Diotsp** [reply]
This is the .raw I used:
[https://files.catbox.moe/oxowna.raw](url) (GitHub does not allow direct uploads in .raw)
It has 3 channels, and is 1100x823.

**Reply by fmw42** [reply]
You can always zip the file and upload the zipped version.

**Reply by Diotsp** [reply]
I didn't think of that; here is the same file in a .zip.
[W32N.zip](https://github.com/ImageMagick/ImageMagick/files/8112757/W32N.zip)

**Comment by fmw42** [comment, marked as answer]
Adobe RAW format appears to be Adobe linear RGB, but your input PNG is non-linear sRGB. You would get a better match if you can find Adobe linear RGB profile and convert that to sRGB using profiles in Imagemagick.  Here is a simple test using -colorspace. I set the colorspace of the RAW image to linear RGB and then convert using -colorspace sRGB

`magick -size 1100x823 -depth 32 -interlace plane -endian msb -define quantum:format=floating-point RGB:x.raw -alpha off -set colorspace RGB -colorspace sRGB -colorspace sRGB y.png
`

My Result:

![y](https://user-images.githubusercontent.com/13458218/155049993-d5a11edc-00c0-4739-b153-eea5a178cb77.png)

Input PNG before converting to RAW in PS:

![154938743-c1e78f2c-1641-44ef-b5cc-eaf6a07133e0](https://user-images.githubusercontent.com/13458218/155050058-547cc941-0a6d-4ce6-ad0b-3a35860ea085.png)

**Reply by Diotsp** [reply]
So to clarify:
My initial problem was due to not specifying `-define quantum:format=floating-point`, as Adobe RAW uses floating-point format (at least for 32bit).
The secondary problem was due to IM not converting the RGB colorspace from the .raw to sRGB, creating the darker appearance. Using `-set colorspace RGB -colorspace sRGB` between the raw input and the output forces that conversion.
Please let me know if I've missed something.

**Reply by fmw42** [reply]
The latter is between linear and non-linear, not raw and sRGB.  Adobe must be saving the raw image as linear RGB. But your png is non-linear sRGB.  If you can find or create a linear RGB compatible with PS saving to raw in linear format, then your could use profiles to convert from Adobe linear RGB to non-linear sRGB. That might give you a better looking result. You can create a linear profile in PS by creating a custom profile with gamma=1. But I do not know what other values to use for the other parameters (white point etc).  See https://community.adobe.com/t5/photoshop-ecosystem-discussions/creating-linear-color-space/td-p/10465287

**Reply by Diotsp** [reply]
I tested when using a RGB (non-sRGB) input, and the `-set colorspace RGB -colorspace sRGB` was still necessary to accurately reproduce the .raw; omitting it caused the result to be still darker. From this I infer that PS compensates for this color shift when displaying a raw.
I will test whether the same is required with 8/16bit files, or with RGBA format as PS supports transparency in its 8/16bit raws.

**Reply by fmw42** [reply]
You probably tested with a non-linear Adobe RGB not a linear Adobe RGB.

**Reply by Diotsp** [reply]
Update: from reading that link, it seems that changing to the 32bit mode automatically uses the linear colorspace (Adobe linear RGB); presumptively this is not true for 8 or 16, which might still use a non-linear colorspace. That would explain why I only encounter this issue when exporting 32bit raws.
It should mean that, since alpha is also supported at those depths, I shouldn't need to convert the colorspace. And after testing, omitting the `define quantum:format` as 16bit doesn't use floating point:
`magick -size 600x718 -depth 16 -interlace none -endian msb RGBA:alphatree.raw alpha.png`
Produces results identical to the source, suggesting 8/16 already use a non-linear colorspace.

**Reply by snibgo** [reply]
Compared to sRGB, AdobeRGB has a similar transfer curve (power 2.2), but the green primary is different, which makes the AdobeSRG gamut larger then the sRGB gamut. If the "raw" file is encoded as AdobeRGB, either linear (no transfer curve) or non-linear (power 2.2) then after reading the raw, the best accuracy is by assigning the appropriate profile, then converting to an sRGB profile.

**Reply by Diotsp** [reply]
> If the "raw" file is encoded as AdobeRGB, either linear (no transfer curve) or non-linear (power 2.2) then after reading the raw, the best accuracy is by assigning the appropriate profile, then converting to an sRGB profile.

So in this case, as from my testing the 32bit mode uses linear, the `-set colorspace RGB` is the appropriate profile, before `-colorspace sRGB` converts?

I further tested my earlier assumption by managing to use an IM command to create a 32bit raw:
`magick W.png -size 1100x823 -depth 32 -interlace none -endian msb -define quantum:format=floating-point -colorspace RGB RGB:32tree.raw`
Which opens correctly in PS, with only very minor difference in appearance from source, perhaps owing to the difference in green primary you mentioned. Note that omitting `-colorspace RGB` produces a significantly brighter result.

**Reply by Diotsp** [reply]
After extended testing, I can confirm that with these options I can use IM to replicate and convert the PS .raw format with a very high degree of accuracy, and achieve the desired behavior.
In short, **the PS 32bits/channel mode uses a floating-point format and a linear colorspace**, which can be compensated for with `-define quantum:format=floating-point` and `-set colorspace RGB -colorspace sRGB` respectively.

Subsequently, I will mark this comment as the answer. Thank you both; this was very helpful.

**Reply by snibgo** [reply]
> So in this case, as from my testing the 32bit mode uses linear, the -set colorspace RGB is the appropriate profile, before -colorspace sRGB converts?

No. If the "raw" file is encoded as AbobeRGB, and you want sRGB, then the most accurate conversion is by using "-profile" instead of "-colorspace".

**Comment by fmw42** [comment]
You should get a better result if you use profiles rather than -set colorspace RGB and -colorspace sRGB
